### PR TITLE
TECS: if height_rate_setpoint is set, only use that one to update altitude reference

### PR DESCRIPTION
**Pull request by sfuhrer**
### Solved Problem
https://github.com/PX4/PX4-Autopilot/pull/21189#issuecomment-1487220075

This reverts the logic to how it was before the TECS refactoring. Not saying it has to stay that way.

### CODE REVIEW

File: src/lib/tecs/TECS.cpp [hunk 1]
```diff
@@ -162,23 +165,32 @@ void TECSReferenceModel::update(const float dt, const AltitudeReferenceState &se
...
-						     param.jerk_max, param.vert_accel_limit, fabsf(delta_trajectory_to_target_m), 0.0f);
+	const float target_climbrate = math::min(param.target_climbrate, param.max_climb_rate);
+	const float target_sinkrate = math::min(param.target_sinkrate, param.max_sink_rate);
 
-		altitude_rate_target = math::constrain(altitude_rate_target, -target_sinkrate, target_climbrate);
+	if (height_rate_controlled) {
+		const float altitude_rate_target = math::constrain(setpoint.alt_rate, -target_sinkrate, target_climbrate);
 
 		_alt_control_traj_generator.updateDurations(altitude_rate_target);
 		_alt_control_traj_generator.updateTraj(dt);
```
Comment (Jaeyoung-Lim) [inline src/lib/tecs/TECS.cpp:175]: Maybe a stupid question, but Why do we still want to update the trajectory generator if we are running height rate control?

**Comment by junwoo091400** [comment]
SITL Tailsitter Ci is suffering attitude failure in roll axis

**Comment by Jaeyoung-Lim** [comment]
@junwoo091400 I dont think this is related to this PR

File: src/lib/tecs/TECS.cpp (hunk 1 above)
Comment (sfuhrer) [reply src/lib/tecs/TECS.cpp:175]: "height rate controlled" is actually the wrong wording. We always control altitude, just that sometimes the interface to TECS is directly the altitude setpoint (eg in Auto flight), and sometimes its the height rate setpoint (ManualPosition, ManualAltitude, Descend).
We should have the discussion next if we can
- only have height rate as interface to TECS
- have the altitude reference model outside of TECS
- have the outer loop for altitude (generate height rate from the current altitude error) outside of TECS

**Comment by sfuhrer** [comment]
Flight tested: https://review.px4.io/plot_app?log=679e33dd-2b0d-4013-baad-31e55fac191a
It now again keeps the current altitude when the pitch stick is centered.
image

File: src/lib/tecs/TECS.cpp (hunk 1 above)
Comment (tstastny) [reply src/lib/tecs/TECS.cpp:175]: > only have height rate as interface to TECS
have the altitude reference model outside of TECS
have the outer loop for altitude (generate height rate from the current altitude error) outside of TECS

yes to all of the above. see comment here: https://github.com/PX4/PX4-Autopilot/pull/21189#issuecomment-1450541349

**Comment by sfuhrer** [comment]
> SITL Tailsitter Ci is suffering attitude failure in roll axis

Seems to have been an intermittent failure, triggered it again and it passed...

**Comment by Jaeyoung-Lim** [comment]
This is from testing offboard sending height rate setpoints
image

**Comment by sfuhrer** [comment]
Updated the PR to revert the logic back to how it was before:
https://github.com/PX4/PX4-Autopilot/blob/1c8ab2a0d7db2d14a6f320ebd8766b5ffaea28fa/src/lib/tecs/TECS.cpp#L455-L483
https://github.com/PX4/PX4-Autopilot/blob/1c8ab2a0d7db2d14a6f320ebd8766b5ffaea28fa/src/lib/tecs/TECS.cpp#L177-L199

That was no retuning of existing frames should be required when switching to the new PX4 release. 
Tested in SITL, seems to work fine. Will try to flight test it tomorrow.
image

**Review by tstastny — APPROVED** [review]
looks correct for the 1.13 logic on code level. (and thanks for the extra clarification of what exactly is happening in each case) - would say this is good once checked out in flight test.

**Comment by sfuhrer** [comment]
Testflown, looked alright, switches from height rate to altitude control in Position mode were smooth: https://review.px4.io/plot_app?log=363600b7-51c6-434f-8e4d-e2b746d59eb9

**Pull request merged**